token-swap/program/src/curve/offset.rs

```diff
@@ -0,0 +1,194 @@
+//! The Uniswap invariant calculator with an extra offset
+
+use crate::{
+    curve::{
+        calculator::{CurveCalculator, DynPack, SwapWithoutFeesResult, TradeDirection},
+        constant_product::swap,
+    },
+    error::SwapError,
+};
+use arrayref::{array_mut_ref, array_ref};
+use solana_program::{
+    program_error::ProgramError,
+    program_pack::{IsInitialized, Pack, Sealed},
+};
+
+/// Offset curve, uses ConstantProduct under the hood, but adds an offset to
+/// one side on swap calculations
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct OffsetCurve {
+    /// Amount to offset the token B liquidity account
+    pub token_b_offset: u64,
+}
+
+impl CurveCalculator for OffsetCurve {
+    /// Constant product swap ensures x * y = constant
+    fn swap_without_fees(
+        &self,
+        source_amount: u128,
+        swap_source_amount: u128,
+        swap_destination_amount: u128,
+        trade_direction: TradeDirection,
+    ) -> Option<SwapWithoutFeesResult> {
+        let token_b_offset = self.token_b_offset as u128;
+        let swap_source_amount = match trade_direction {
+            TradeDirection::AtoB => swap_source_amount,
+            TradeDirection::BtoA => swap_source_amount.checked_add(token_b_offset)?,
+        };
+        let swap_destination_amount = match trade_direction {
+            TradeDirection::AtoB => swap_destination_amount.checked_add(token_b_offset)?,
+            TradeDirection::BtoA => swap_destination_amount,
+        };
+        swap(source_amount, swap_source_amount, swap_destination_amount)
+    }
+
+    fn validate(&self) -> Result<(), SwapError> {
+        if self.token_b_offset == 0 {
```

Comment:
Technically this is fine, but people should just make a normal constant product curve if they don't want any offset